sklearn/gaussian_process/kernels.py

```diff
@@ -1787,3 +1787,182 @@ def is_stationary(self):
+         `n_features` marked as True for an active index.
+         If `active_dims` is an integer array, then it should be size of
+         `n_active_dims` denoting the indices of the active dimensions.
+
+    """
+
+    def __init__(self, kernel, active_dims):
+        self.active_dims = active_dims
+        self.kernel = kernel
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """Return the kernel k(X, Y) and optionally its gradient on selected
+           dimensions.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples_X, n_features)
+            Left argument of the returned kernel k(X, Y)
+
+        Y : array, shape (n_samples_Y, n_features), (optional, default=None)
+            Right argument of the returned kernel k(X, Y). If None, k(X, X)
+            evaluated instead.
+
+        eval_gradient : bool (optional, default=False)
+            Determines whether the gradient with respect to the kernel
+            hyperparameter is determined. Only supported when Y is None.
+
+        Returns
+        -------
+        K : array, shape (n_samples_X, n_samples_Y)
+            Kernel k(X[:,active_dims], Y[:,active_dims])
+
+        K_gradient : array (opt.) shape (n_samples X n_samples X active_dims)
+            The gradient of the kernel k(X, X) with respect to the
+            hyperparameter of the kernel. Only returned when eval_gradient
+            is True.
+        """
+
+        active_dims = np.asarray(self.active_dims)
+
+        if (np.issubdtype(active_dims.dtype, np.bool) &
+           (active_dims.shape == X.shape[1])):
+            active_dims = np.where(active_dims)[0]
+        else:
+            active_dims = active_dims.astype(int)
+        if active_dims.size == 0:
+            raise ValueError("active_dims should be have at least one \
+                              element, current size: %d " % active_dims.size)
+
+        return self.kernel_(X[:, active_dims],
+                            None if Y is None else Y[:, active_dims],
+                            eval_gradient)
+
+    def get_params(self, deep=True):
+        """Get parameters of this kernel.
+
+        Parameters
+        ----------
+        deep: boolean, optional
+            If True, will return of string of parameter names
+            mapped to their corresponding values.
+
+        Returns
+        -------
+        params : dict mapping of string to any
+            Parameter names mapped to their values.
+        """
+
+        params = dict(kernel=self.kernel_, active_dims=self.active_dims)
+        if deep:
+            deep_items = self.kernel_.get_params().items()
+            params.update(('kernel__' + k, val) for k, val in deep_items)
+        return params
+
+    @property
+    def hyperparameters(self):
+        """Returns a list of all hyperparameters."""
+        r = []
+        for hyperparameter in self.kernel_.hyperparameters:
```

Comment:
This might be problematic if we have an anisotropic kernel (RBF or Matern) and select only some of its dimensions because some of the hyperparameters (entries of length-scales for deactivated dimensions) then do not actually affect the kernel. I am not sure what to do about it. We could try to blend out these hyperparameter entries (and also the corresponding theta dimensions) but I am not sure if this does not result in new problems. What do you think?